### ISSUE TITLE
Option to step the circuit periodically even in the absence of input

### ISSUE THREAD

**Issue by mihaibudiu** [feldera/feldera#2319]
This is useful for programs that use "now()". If there is no data, such programs will do nothing. Inserting empty inputs will trigger a re-evaluation of the now() function.

Technically only one input adapter has to do that for any program.

**Comment by blp** [comment]
If an empty input means one with no records, then this won't do anything useful because a step is triggered by a threshold on the number of buffered input records.

**Comment by mihaibudiu** [comment]
The step should be triggered, that's what I want.

**Comment by blp** [comment]
> The step should be triggered, that's what I want.

OK. That's fine, it's just not related to the input adapters. It would be an additional step trigger (">=N msecs elapsed since the last step").

**Comment by mihaibudiu** [comment]
Well, the adapters will have to pull the trigger. Or would it be the pipeline manager?

**Comment by mihaibudiu** [comment]
Feel free to retag the issue.

**Comment by blp** [comment]
> Well, the adapters will have to pull the trigger. Or would it be the pipeline manager?

It's the controller inside the pipeline manager. An additional disjunction in this check inside `crates/adapters/src/controller/mod.rs`:

```rust
                    // We have sufficient buffered inputs or the buffering delay has expired or
                    // the client explicitly requested the circuit to run -- kick the circuit to
                    // consume buffered data.
                    // Use strict inequality in case `min_batch_size_records` is 0.
                    if controller.status.step_requested()
                        || buffered_records > min_batch_size_records
                        || start
                            .map(|start| start.elapsed() >= max_buffering_delay)
                            .unwrap_or(false)
```

**Comment by ryzhyk** [comment]
The `adapters` crate already has a way to force a step on the circuit. We just need a config option to do it periodically at a predefined interval.

**Comment by ryzhyk** [comment]
Or we can just to it every second

**Comment by mihaibudiu** [comment]
Yes, one question is where the config option is too. Adapter options (also called "connectors", we should settle on a terminology) are very visible in the SQL program, whereas the other ones are hidden.

**Comment by ryzhyk** [comment]
Implemented in #2336

**Comment by mihaibudiu** [comment]
Can we close this?
